// @ironsoftware/ironpdf/src/public/affix.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AffixFonts = void 0;
/**
 * Supported Affix Fonts
 */
var AffixFonts;
(function (AffixFonts) {
    AffixFonts[AffixFonts["Arial"] = 0] = "Arial";
    AffixFonts[AffixFonts["Arial_Bold"] = 1] = "Arial_Bold";
    AffixFonts[AffixFonts["Arial_BoldItalic"] = 2] = "Arial_BoldItalic";
    AffixFonts[AffixFonts["Arial_Italic"] = 3] = "Arial_Italic";
    AffixFonts[AffixFonts["Courier"] = 4] = "Courier";
    AffixFonts[AffixFonts["Courier_BoldOblique"] = 5] = "Courier_BoldOblique";
    AffixFonts[AffixFonts["Courier_Oblique"] = 6] = "Courier_Oblique";
    AffixFonts[AffixFonts["Courier_Bold"] = 7] = "Courier_Bold";
    AffixFonts[AffixFonts["CourierNew"] = 8] = "CourierNew";
    AffixFonts[AffixFonts["CourierNew_Bold"] = 9] = "CourierNew_Bold";
    AffixFonts[AffixFonts["CourierNew_BoldItalic"] = 10] = "CourierNew_BoldItalic";
    AffixFonts[AffixFonts["CourierNew_Italic"] = 11] = "CourierNew_Italic";
    AffixFonts[AffixFonts["Helvetica"] = 12] = "Helvetica";
    AffixFonts[AffixFonts["Helvetica_Bold"] = 13] = "Helvetica_Bold";
    AffixFonts[AffixFonts["Helvetica_BoldOblique"] = 14] = "Helvetica_BoldOblique";
    AffixFonts[AffixFonts["Helvetica_Oblique"] = 15] = "Helvetica_Oblique";
    AffixFonts[AffixFonts["Symbol"] = 16] = "Symbol";
    AffixFonts[AffixFonts["TimesNewRoman"] = 17] = "TimesNewRoman";
    AffixFonts[AffixFonts["TimesNewRoman_Bold"] = 18] = "TimesNewRoman_Bold";
    AffixFonts[AffixFonts["TimesNewRoman_BoldItalic"] = 19] = "TimesNewRoman_BoldItalic";
    AffixFonts[AffixFonts["TimesNewRoman_Italic"] = 20] = "TimesNewRoman_Italic";
    AffixFonts[AffixFonts["ZapfDingbats"] = 21] = "ZapfDingbats";
})(AffixFonts = exports.AffixFonts || (exports.AffixFonts = {}));
