// @ironsoftware/ironpdf/src/public/image.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImageBehavior = exports.ImageType = void 0;
/**
 * Supported image file type
 */
var ImageType;
(function (ImageType) {
    ImageType[ImageType["PNG"] = 0] = "PNG";
    ImageType[ImageType["JPG"] = 1] = "JPG";
    ImageType[ImageType["BMP"] = 2] = "BMP";
})(ImageType = exports.ImageType || (exports.ImageType = {}));
/**
 * Describes how image should be placed on the PDF page
 */
var ImageBehavior;
(function (ImageBehavior) {
    /**
     * Image should be placed on center of the page.
     */
    ImageBehavior[ImageBehavior["CenteredOnPage"] = 0] = "CenteredOnPage";
    /**
     * Image should fit to the page.
     */
    ImageBehavior[ImageBehavior["FitToPage"] = 1] = "FitToPage";
    /**
     * Image should fit to the page and keep aspect ratio.
     */
    ImageBehavior[ImageBehavior["FitToPageAndMaintainAspectRatio"] = 2] = "FitToPageAndMaintainAspectRatio";
    /**
     * Page should fit to the image.
     */
    ImageBehavior[ImageBehavior["CropPage"] = 3] = "CropPage";
    /**
     * Image should be placed to the left top corner of the page.
     */
    ImageBehavior[ImageBehavior["TopLeftCornerOfPage"] = 4] = "TopLeftCornerOfPage";
    /**
     * Image should be placed to the left bottom corner of the page.
     */
    ImageBehavior[ImageBehavior["BottomLeftCornerOfPage"] = 5] = "BottomLeftCornerOfPage";
    /**
     * Image should be placed to the right top corner of the page.
     */
    ImageBehavior[ImageBehavior["TopRightCornerOfPage"] = 6] = "TopRightCornerOfPage";
    /**
     * Image should be placed to the right bottom corner of the page.
     */
    ImageBehavior[ImageBehavior["BottomRightCornerOfPage"] = 7] = "BottomRightCornerOfPage";
})(ImageBehavior = exports.ImageBehavior || (exports.ImageBehavior = {}));
