// @ironsoftware/ironpdf/src/public/paper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaperSizeUnit = exports.PdfPaperOrientation = exports.PaperSize = void 0;
/**
 * Defines the target virtual paper size the PDF.  Relates to real-world paper-sizes.
 */
var PaperSize;
(function (PaperSize) {
    /**
     * 8.5in x 11in
     */
    PaperSize[PaperSize["Letter"] = 0] = "Letter";
    /**
     * 8.5in x 11in
     */
    PaperSize[PaperSize["Legal"] = 1] = "Legal";
    /**
     * 210mm x 297mm
     */
    PaperSize[PaperSize["A4"] = 2] = "A4";
    /**
     * 17in x 22in
     */
    PaperSize[PaperSize["CSheet"] = 3] = "CSheet";
    /**
     * 22in x 34in
     */
    PaperSize[PaperSize["DSheet"] = 4] = "DSheet";
    /**
     * 34in x 44in
     */
    PaperSize[PaperSize["ESheet"] = 5] = "ESheet";
    /**
     * 8.5in x 11in
     */
    PaperSize[PaperSize["LetterSmall"] = 6] = "LetterSmall";
    /**
     * 11in x 17in
     */
    PaperSize[PaperSize["Tabloid"] = 7] = "Tabloid";
    /**
     * 17in x 11in
     */
    PaperSize[PaperSize["Ledger"] = 8] = "Ledger";
    /**
     * 5.5in x 8.5in
     */
    PaperSize[PaperSize["Statement"] = 9] = "Statement";
    /**
     * 7.25in x 10.5in
     */
    PaperSize[PaperSize["Executive"] = 10] = "Executive";
    /**
     * 297mm x 420mm
     */
    PaperSize[PaperSize["A3"] = 11] = "A3";
    /**
     * 210mm x 297mm
     */
    PaperSize[PaperSize["A4Small"] = 12] = "A4Small";
    /**
     * 148mm x 210mm
     */
    PaperSize[PaperSize["A5"] = 13] = "A5";
    /**
     * 250mm x 353mm
     */
    PaperSize[PaperSize["B4"] = 14] = "B4";
    /**
     * 176mm x 250mm
     */
    PaperSize[PaperSize["B5"] = 15] = "B5";
    /**
     * 8.5in x 13in
     */
    PaperSize[PaperSize["Folio"] = 16] = "Folio";
    /**
     * 215mm x 275mm
     */
    PaperSize[PaperSize["Quarto"] = 17] = "Quarto";
    /**
     * 10in x 14in
     */
    PaperSize[PaperSize["Standard10x14"] = 18] = "Standard10x14";
    /**
     * 11in x 17in
     */
    PaperSize[PaperSize["Standard11x17"] = 19] = "Standard11x17";
    /**
     * 8.5in x 11in
     */
    PaperSize[PaperSize["Note"] = 20] = "Note";
    /**
     * 3.875in x 8.875in
     */
    PaperSize[PaperSize["Number9Envelope"] = 21] = "Number9Envelope";
    /**
     * 4.125in x 9.5in
     */
    PaperSize[PaperSize["Number10Envelope"] = 22] = "Number10Envelope";
    /**
     * 4.5in x 10.375in
     */
    PaperSize[PaperSize["Number11Envelope"] = 23] = "Number11Envelope";
    /**
     * 4.75in x 11in
     */
    PaperSize[PaperSize["Number12Envelope"] = 24] = "Number12Envelope";
    /**
     * 5in x 11.5in
     */
    PaperSize[PaperSize["Number14Envelope"] = 25] = "Number14Envelope";
    /**
     * 110mm x 220mm
     */
    PaperSize[PaperSize["DLEnvelope"] = 26] = "DLEnvelope";
    /**
     * 162mm x 229mm
     */
    PaperSize[PaperSize["C5Envelope"] = 27] = "C5Envelope";
    /**
     * 324mm x 458mm
     */
    PaperSize[PaperSize["C3Envelope"] = 28] = "C3Envelope";
    /**
     * 229mm x 324mm
     */
    PaperSize[PaperSize["C4Envelope"] = 29] = "C4Envelope";
    /**
     * 114mm x 162mm
     */
    PaperSize[PaperSize["C6Envelope"] = 30] = "C6Envelope";
    /**
     * 114mm x 229mm
     */
    PaperSize[PaperSize["C65Envelope"] = 31] = "C65Envelope";
    /**
     * 250mm x 353mm
     */
    PaperSize[PaperSize["B4Envelope"] = 32] = "B4Envelope";
    /**
     * 176mm x 250mm
     */
    PaperSize[PaperSize["B5Envelope"] = 33] = "B5Envelope";
    /**
     * 176mm x 125mm
     */
    PaperSize[PaperSize["B6Envelope"] = 34] = "B6Envelope";
    /**
     * 110mm x 230mm
     */
    PaperSize[PaperSize["ItalyEnvelope"] = 35] = "ItalyEnvelope";
    /**
     * 3.875in x 7.5in
     */
    PaperSize[PaperSize["MonarchEnvelope"] = 36] = "MonarchEnvelope";
    /**
     * 3.625in x 6.5in
     */
    PaperSize[PaperSize["PersonalEnvelope"] = 37] = "PersonalEnvelope";
    /**
     * 14.875in x 11in
     */
    PaperSize[PaperSize["USStandardFanfold"] = 38] = "USStandardFanfold";
    /**
     * 8.5in x 12in
     */
    PaperSize[PaperSize["GermanStandardFanfold"] = 39] = "GermanStandardFanfold";
    /**
     * 8.5in x 13in
     */
    PaperSize[PaperSize["GermanLegalFanfold"] = 40] = "GermanLegalFanfold";
    /**
     * 250mm x 353mm
     */
    PaperSize[PaperSize["IsoB4"] = 41] = "IsoB4";
    /**
     * 100mm x 148mm
     */
    PaperSize[PaperSize["JapanesePostcard"] = 42] = "JapanesePostcard";
    /**
     * 9in x 11in
     */
    PaperSize[PaperSize["Standard9x11"] = 43] = "Standard9x11";
    /**
     * 10in x 11in
     */
    PaperSize[PaperSize["Standard10x11"] = 44] = "Standard10x11";
    /**
     * 15in x 11in
     */
    PaperSize[PaperSize["Standard15x11"] = 45] = "Standard15x11";
    /**
     * 220mm x 220mm
     */
    PaperSize[PaperSize["InviteEnvelope"] = 46] = "InviteEnvelope";
    /**
     * 9.275in x 12in
     */
    PaperSize[PaperSize["LetterExtra"] = 47] = "LetterExtra";
    /**
     * 9.275in x 15in
     */
    PaperSize[PaperSize["LegalExtra"] = 48] = "LegalExtra";
    /**
     * 11.69in x 18in
     */
    PaperSize[PaperSize["TabloidExtra"] = 49] = "TabloidExtra";
    /**
     * 236mm x 322mm
     */
    PaperSize[PaperSize["A4Extra"] = 50] = "A4Extra";
    /**
     * 8.275in x 11in
     */
    PaperSize[PaperSize["LetterTransverse"] = 51] = "LetterTransverse";
    /**
     * 210mm x 297mm
     */
    PaperSize[PaperSize["A4Transverse"] = 52] = "A4Transverse";
    /**
     * 9.275in x 12in
     */
    PaperSize[PaperSize["LetterExtraTransverse"] = 53] = "LetterExtraTransverse";
    /**
     * 227mm x 356mm
     */
    PaperSize[PaperSize["APlus"] = 54] = "APlus";
    /**
     * 305mm x 487mm
     */
    PaperSize[PaperSize["BPlus"] = 55] = "BPlus";
    /**
     * 8.5in x 12.69in
     */
    PaperSize[PaperSize["LetterPlus"] = 56] = "LetterPlus";
    /**
     * 210mm x 330mm
     */
    PaperSize[PaperSize["A4Plus"] = 57] = "A4Plus";
    /**
     * 148mm x 210mm
     */
    PaperSize[PaperSize["A5Transverse"] = 58] = "A5Transverse";
    /**
     * 182mm x 257mm
     */
    PaperSize[PaperSize["B5Transverse"] = 59] = "B5Transverse";
    /**
     * 322mm x 445mm
     */
    PaperSize[PaperSize["A3Extra"] = 60] = "A3Extra";
    /**
     * 174mm x 235mm
     */
    PaperSize[PaperSize["A5Extra"] = 61] = "A5Extra";
    /**
     * 201mm x 276mm
     */
    PaperSize[PaperSize["B5Extra"] = 62] = "B5Extra";
    /**
     * 420mm x 594mm
     */
    PaperSize[PaperSize["A2"] = 63] = "A2";
    /**
     * 297mm x 420mm
     */
    PaperSize[PaperSize["A3Transverse"] = 64] = "A3Transverse";
    /**
     * 322mm x 445mm
     */
    PaperSize[PaperSize["A3ExtraTransverse"] = 65] = "A3ExtraTransverse";
    /**
     * 200mm x 148mm
     */
    PaperSize[PaperSize["JapaneseDoublePostcard"] = 66] = "JapaneseDoublePostcard";
    /**
     * 105mm x 148mm
     */
    PaperSize[PaperSize["A6"] = 67] = "A6";
    /**
     * 11in x 8.5in
     */
    PaperSize[PaperSize["LetterRotated"] = 68] = "LetterRotated";
    /**
     * 420mm x 297mm
     */
    PaperSize[PaperSize["A3Rotated"] = 69] = "A3Rotated";
    /**
     * 297mm x 210mm
     */
    PaperSize[PaperSize["A4Rotated"] = 70] = "A4Rotated";
    /**
     * 210mm x 148mm
     */
    PaperSize[PaperSize["A5Rotated"] = 71] = "A5Rotated";
    /**
     * 364mm x 257mm
     */
    PaperSize[PaperSize["B4JisRotated"] = 72] = "B4JisRotated";
    /**
     * 257mm x 182mm
     */
    PaperSize[PaperSize["B5JisRotated"] = 73] = "B5JisRotated";
    /**
     * 148mm x 100mm
     */
    PaperSize[PaperSize["JapanesePostcardRotated"] = 74] = "JapanesePostcardRotated";
    /**
     * 148mm x 200mm
     */
    PaperSize[PaperSize["JapaneseDoublePostcardRotated"] = 75] = "JapaneseDoublePostcardRotated";
    /**
     * 148mm x 105mm
     */
    PaperSize[PaperSize["A6Rotated"] = 76] = "A6Rotated";
    /**
     * 128mm x 182mm
     */
    PaperSize[PaperSize["B6Jis"] = 77] = "B6Jis";
    /**
     * 182mm x 128mm
     */
    PaperSize[PaperSize["B6JisRotated"] = 78] = "B6JisRotated";
    /**
     * 12in x 11in
     */
    PaperSize[PaperSize["Standard12x11"] = 79] = "Standard12x11";
    /**
     * 146mm x 215mm
     */
    PaperSize[PaperSize["Prc16K"] = 80] = "Prc16K";
    /**
     * 97mm x 151mm
     */
    PaperSize[PaperSize["Prc32K"] = 81] = "Prc32K";
    /**
     * 97mm x 151mm
     */
    PaperSize[PaperSize["Prc32KBig"] = 82] = "Prc32KBig";
    /**
     * 102mm x 165mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber1"] = 83] = "PrcEnvelopeNumber1";
    /**
     * 102mm x 176mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber2"] = 84] = "PrcEnvelopeNumber2";
    /**
     * 125mm x 176mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber3"] = 85] = "PrcEnvelopeNumber3";
    /**
     * 110mm x 208mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber4"] = 86] = "PrcEnvelopeNumber4";
    /**
     * 110mm x 220mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber5"] = 87] = "PrcEnvelopeNumber5";
    /**
     * 120mm x 230mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber6"] = 88] = "PrcEnvelopeNumber6";
    /**
     * 160mm x 230mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber7"] = 89] = "PrcEnvelopeNumber7";
    /**
     * 120mm x 309mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber8"] = 90] = "PrcEnvelopeNumber8";
    /**
     * 229mm x 324mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber9"] = 91] = "PrcEnvelopeNumber9";
    /**
     * 324mm x 458mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber10"] = 92] = "PrcEnvelopeNumber10";
    /**
     * 146mm x 215mm
     */
    PaperSize[PaperSize["Prc16KRotated"] = 93] = "Prc16KRotated";
    /**
     * 97mm x 151mm
     */
    PaperSize[PaperSize["Prc32KRotated"] = 94] = "Prc32KRotated";
    /**
     * 97mm x 151mm
     */
    PaperSize[PaperSize["Prc32KBigRotated"] = 95] = "Prc32KBigRotated";
    /**
     * 165mm x 102mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber1Rotated"] = 96] = "PrcEnvelopeNumber1Rotated";
    /**
     * 176mm x 102mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber2Rotated"] = 97] = "PrcEnvelopeNumber2Rotated";
    /**
     * 176mm x 125mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber3Rotated"] = 98] = "PrcEnvelopeNumber3Rotated";
    /**
     * 208mm x 110mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber4Rotated"] = 99] = "PrcEnvelopeNumber4Rotated";
    /**
     * 220mm x 110mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber5Rotated"] = 100] = "PrcEnvelopeNumber5Rotated";
    /**
     * 230mm x 120mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber6Rotated"] = 101] = "PrcEnvelopeNumber6Rotated";
    /**
     * 230mm x 160mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber7Rotated"] = 102] = "PrcEnvelopeNumber7Rotated";
    /**
     * 309mm x 120mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber8Rotated"] = 103] = "PrcEnvelopeNumber8Rotated";
    /**
     * 324mm x 229mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber9Rotated"] = 104] = "PrcEnvelopeNumber9Rotated";
    /**
     * 458mm x 324mm
     */
    PaperSize[PaperSize["PrcEnvelopeNumber10Rotated"] = 105] = "PrcEnvelopeNumber10Rotated";
})(PaperSize = exports.PaperSize || (exports.PaperSize = {}));
/**
 * Paper Orientation
 */
var PdfPaperOrientation;
(function (PdfPaperOrientation) {
    /**
     * Paper is oriented vertically
     */
    PdfPaperOrientation[PdfPaperOrientation["Portrait"] = 0] = "Portrait";
    /**
     * Paper is oriented horizontally
     */
    PdfPaperOrientation[PdfPaperOrientation["Landscape"] = 1] = "Landscape";
})(PdfPaperOrientation = exports.PdfPaperOrientation || (exports.PdfPaperOrientation = {}));
/**
 * Dimensions or unit of mature
 */
var PaperSizeUnit;
(function (PaperSizeUnit) {
    PaperSizeUnit[PaperSizeUnit["Millimeter"] = 0] = "Millimeter";
    PaperSizeUnit[PaperSizeUnit["Inch"] = 1] = "Inch";
    PaperSizeUnit[PaperSizeUnit["Centimeter"] = 2] = "Centimeter";
    PaperSizeUnit[PaperSizeUnit["Points"] = 3] = "Points";
})(PaperSizeUnit = exports.PaperSizeUnit || (exports.PaperSizeUnit = {}));
