// @ironsoftware/ironpdf/src/internal/zod/affixSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.textAffixSchema = exports.affixFontsSchema = exports.htmlAffixSchema = void 0;
// Generated by ts-to-zod
const zod_1 = require("zod");
const affix_1 = require("../../public/affix");
exports.htmlAffixSchema = zod_1.z.object({
    htmlFragment: zod_1.z.string(),
    maxHeight: zod_1.z.number().optional(),
    dividerLine: zod_1.z
        .boolean()
        .optional(),
    dividerLineColor: zod_1.z
        .string()
        .optional(),
    loadStylesAndCSSFromMainHtmlDocument: zod_1.z
        .boolean()
        .optional()
});
exports.affixFontsSchema = zod_1.z.nativeEnum(affix_1.AffixFonts);
exports.textAffixSchema = zod_1.z.object({
    leftText: zod_1.z.string().optional(),
    centerText: zod_1.z.string().optional(),
    rightText: zod_1.z.string().optional(),
    dividerLine: zod_1.z
        .boolean()
        .optional(),
    dividerLineColor: zod_1.z
        .string()
        .optional(),
    font: exports.affixFontsSchema.optional(),
    fontSize: zod_1.z
        .number()
        .optional()
});
