// @ironsoftware/ironpdf/src/internal/zod/signatureSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.digitalSignatureSchema = exports.signatureImageSchema = void 0;
const zod_1 = require("zod");
const typeSchema_1 = require("./typeSchema");
exports.signatureImageSchema = zod_1.z.object({
    SignatureImagePath: zod_1.z.string().optional(),
    SignatureImageBuffer: typeSchema_1.bufferSchema.optional(),
    SignatureImagePosition: typeSchema_1.cropRectangleSchema.optional(),
    SignatureImagePageIndex: zod_1.z.number().optional()
});
exports.digitalSignatureSchema = zod_1.z.object({
    timeStampUrl: zod_1.z.string().optional(),
    certificatePath: zod_1.z.string().optional(),
    certificateBuffer: typeSchema_1.bufferSchema.optional(),
    certificatePassword: zod_1.z.string().optional(),
    signingReason: zod_1.z.string().optional(),
    signingLocation: zod_1.z.string().optional(),
    signatureDate: typeSchema_1.dateSchema.optional(),
    signatureImage: exports.signatureImageSchema.optional()
});
