// @ironsoftware/ironpdf/src/internal/zod/typeSchema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapStringSchema = exports.booleanSchema = exports.numberSchema = exports.bufferArraySchema = exports.stringArraySchema = exports.barcodeTypeSchema = exports.dateSchema = exports.pdfInputSchema = exports.changeTrackingModesSchema = exports.chromeGpuModesSchema = exports.pdfPasswordSchema = exports.useMarginsSchema = exports.imageBufferSchema = exports.bufferSchema = exports.imageFilePathSchema = exports.urlSchema = exports.stringSchema = exports.urlStringSchema = exports.htmlFilePathSchema = exports.htmlStringSchema = exports.zipFilePathSchema = exports.pdfFilePathSchema = exports.filePathSchema = exports.marginConfigSchema = exports.cropRectangleSchema = exports.lengthSchema = exports.saveOptionsSchema = exports.measurementUnitSchema = exports.pdfPageSelectionSchema = exports.fitToPaperModesSchema = exports.cssMediaTypeSchema = exports.verticalAlignmentSchema = exports.horizontalAlignmentSchema = void 0;
const zod_1 = require("zod");
const types_1 = require("../../public/types");
const pdfDocumentSchema_1 = require("./pdfDocumentSchema");
const buffer_1 = require("buffer");
exports.horizontalAlignmentSchema = zod_1.z.nativeEnum(types_1.HorizontalAlignment);
exports.verticalAlignmentSchema = zod_1.z.nativeEnum(types_1.VerticalAlignment);
exports.cssMediaTypeSchema = zod_1.z.nativeEnum(types_1.CssMediaType);
exports.fitToPaperModesSchema = zod_1.z.nativeEnum(types_1.FitToPaperModes);
exports.pdfPageSelectionSchema = zod_1.z.union([
    zod_1.z.number(),
    zod_1.z.array(zod_1.z.number()),
    zod_1.z.literal("all"),
    zod_1.z.undefined()
]);
exports.measurementUnitSchema = zod_1.z.nativeEnum(types_1.MeasurementUnit);
exports.saveOptionsSchema = zod_1.z.object({
    userPassword: zod_1.z.string().optional(),
    ownerPassword: zod_1.z
        .string()
        .optional(),
    saveAsPdfA: zod_1.z
        .boolean()
        .optional(),
    incremental: zod_1.z
        .boolean()
        .optional()
});
exports.lengthSchema = zod_1.z.object({
    value: zod_1.z.number(),
    unit: exports.measurementUnitSchema
});
exports.cropRectangleSchema = zod_1.z.object({
    x: zod_1.z.number().optional(),
    y: zod_1.z.number().optional(),
    width: zod_1.z.number().optional(),
    height: zod_1.z.number().optional()
});
exports.marginConfigSchema = zod_1.z.object({
    default: zod_1.z.number().optional(),
    top: zod_1.z.number().optional(),
    right: zod_1.z.number().optional(),
    bottom: zod_1.z.number().optional(),
    left: zod_1.z.number().optional()
});
exports.filePathSchema = zod_1.z.string();
exports.pdfFilePathSchema = zod_1.z.string();
exports.zipFilePathSchema = zod_1.z.string();
exports.htmlStringSchema = zod_1.z.string();
exports.htmlFilePathSchema = zod_1.z.string();
exports.urlStringSchema = zod_1.z.string();
exports.stringSchema = zod_1.z.string();
exports.urlSchema = zod_1.z.instanceof(URL);
exports.imageFilePathSchema = zod_1.z.string();
exports.bufferSchema = zod_1.z.instanceof(buffer_1.Buffer);
exports.imageBufferSchema = exports.bufferSchema;
exports.useMarginsSchema = zod_1.z.nativeEnum(types_1.UseMargins);
exports.pdfPasswordSchema = zod_1.z.object({
    userPassword: zod_1.z.string().optional(),
    ownerPassword: zod_1.z.string().optional()
});
exports.chromeGpuModesSchema = zod_1.z.nativeEnum(types_1.ChromeGpuModes);
exports.changeTrackingModesSchema = zod_1.z.nativeEnum(types_1.ChangeTrackingModes);
exports.pdfInputSchema = zod_1.z.union([pdfDocumentSchema_1.pdfDocumentSchema, exports.bufferSchema, exports.htmlStringSchema, exports.htmlFilePathSchema, exports.zipFilePathSchema, exports.pdfFilePathSchema, exports.urlSchema, exports.urlStringSchema]);
exports.dateSchema = zod_1.z.instanceof(Date);
exports.barcodeTypeSchema = zod_1.z.nativeEnum(types_1.BarcodeType);
exports.stringArraySchema = zod_1.z.array(zod_1.z.string());
exports.bufferArraySchema = zod_1.z.array(exports.bufferSchema);
exports.numberSchema = zod_1.z.number();
exports.booleanSchema = zod_1.z.boolean();
exports.mapStringSchema = zod_1.z.map(zod_1.z.string(), zod_1.z.string());
